synnax/pkg/distribution/framer/encoder/encoder.go

```diff
@@ -0,0 +1,183 @@
+		index += 1
+	}
+
+	// First byte for any is
+	// sameData_strongAligned_allChannels_curDataSize (5 bits)
+	returnArray = append(returnArray, byte((sizeFlag<<2)|(alignFlag<<1)|(channelFlag)))
+
+	if sizeFlag == 1 {
+		bytesToAppend := make([]byte, 4)
+		binary.LittleEndian.PutUint32(bytesToAppend, uint32(curDataSize))
+		returnArray = append(returnArray, bytesToAppend...)
+	}
+	if alignFlag == 1 {
+		timeRangeAppend := make([]byte, 8)
+		binary.LittleEndian.PutUint64(timeRangeAppend, uint64(src.Series[index-1].TimeRange.Start))
+		returnArray = append(returnArray, timeRangeAppend...)
+		binary.LittleEndian.PutUint64(timeRangeAppend, uint64(src.Series[index-1].TimeRange.End))
+		returnArray = append(returnArray, timeRangeAppend...)
+	}
+
+	index = 0
+
+	for i := 0; i < len(m.keys); i++ {
+		if index >= len(src.Keys) || src.Keys[index] != m.keys[i] {
+			continue
+		}
+		lenSeriesData := uint32(len(src.Series[index].Data))
+		dataSize := uint32(m.dtypes[i].Density())
+		// Adding Key
+		if channelFlag == 0 {
+			keyInformation := make([]byte, 4)
+			binary.LittleEndian.PutUint32(keyInformation, uint32(src.Keys[index]))
+			returnArray = append(returnArray, keyInformation...)
+		}
+		// Adding Data Length
+		if sizeFlag == 0 {
+			bytesToAppend := make([]byte, 4)
+			binary.LittleEndian.PutUint32(bytesToAppend, uint32(lenSeriesData/dataSize))
+			returnArray = append(returnArray, bytesToAppend...)
+		}
+		// Adding Data
+		for j := 0; j < int(lenSeriesData); j++ {
+			returnArray = append(returnArray, src.Series[index].Data[j])
+		}
+		// Adding Time range
+		if alignFlag == 0 {
+			timeRangeAppend := make([]byte, 8)
+			binary.LittleEndian.PutUint64(timeRangeAppend, uint64(src.Series[index].TimeRange.Start))
+			returnArray = append(returnArray, timeRangeAppend...)
+			binary.LittleEndian.PutUint64(timeRangeAppend, uint64(src.Series[index].TimeRange.End))
+			returnArray = append(returnArray, timeRangeAppend...)
+		}
+
+		index += 1
+	}
+	return returnArray, nil
+}
+
+func (m EncoderDecoder) Decode(src []byte) (dst framer.Frame, err error) {
+	var (
+		returnStruct                                   = framer.Frame{}
+		sizeFlag, alignFlag, channelFlag, index int    = 0, 0, 0, 0
+		sizeRepresentation, curSize             uint32 = 0, 0
+		timeRangeStart, timeRangeEnd            uint64 = 0, 0
+	)
+
+	sizeFlag = int((src[index] >> 2) & 1)
+	alignFlag = int((src[index] >> 1) & 1)
+	channelFlag = int(src[index] & 1)
+	index += 1
+
+	if sizeFlag == 1 {
+		sizeRepresentation = binary.LittleEndian.Uint32(src[index:])
+		index += 4
+	}
+
+	if alignFlag == 1 {
+		timeRangeStart = binary.LittleEndian.Uint64(src[index:])
+		timeRangeEnd = binary.LittleEndian.Uint64(src[index+8:])
+		index += 16
+	}
+
+	if channelFlag == 1 {
+		returnStruct.Keys = m.keys
+	}
+
+	for i := 0; i < len(m.keys); i++ {
+		if channelFlag == 0 {
+			if index < len(src) && channel.Key(binary.LittleEndian.Uint32(src[index:])) == m.keys[i] {
```

Comment:
If statement could be better structured. Just check for the converse and use continue? (or is it inverse). Anyway, you don't need an else statement here.